relayer/processor/path_end_runtime.go

```diff
@@ -290,6 +290,40 @@ func (pathEnd *pathEndRuntime) shouldTerminate(ibcMessagesCache IBCMessagesCache
 	return false
 }
 
+// checkForMisbehaviour is called for each attempt to update the light client on this path end. The proposed header will
+// be compared against the cached trusted header for the same block height to determine if there is a deviation in the
+// consensus states, if there is no cached trusted header then it will be queried from the counterparty further down in
+// the call stack. If a deviation is found a MsgSubmitMisbehaviour will be composed and broadcasted to freeze the
+// light client. If misbehaviour is detected true will be returned and the pathEndRuntime should terminate.
+// If no misbehaviour is detected false will be returned along with a nil error.
+func (pathEnd *pathEndRuntime) checkForMisbehaviour(
+	ctx context.Context,
+	state provider.ClientState,
+	counterparty *pathEndRuntime,
+) (bool, error) {
+	cachedHeader := counterparty.ibcHeaderCache[state.ConsensusHeight.RevisionHeight]
+
+	misbehaviour, err := provider.CheckForMisbehaviour(ctx, pathEnd.chainProvider.Codec(), counterparty.chainProvider, pathEnd.info.ClientID, state.Header, cachedHeader)
```

Comment:
Rather than adding a `Codec()` provider function, since we only need the tm client codec for this, maybe we can have this be a minimal codec within the provider or pathEnd

Reply:
Great idea! I'll get something implemented inside of provider since that's where we are using it anyways

Reply:
There isn't a great way to bring this into the provider without either instantiating a new codec on the fly or using a global singleton so I just added this to the `pathEndRuntime`